### DISCUSSION TITLE
Waterfall Chart

### DISCUSSION THREAD

**Discussion by hughess** [evidence-dev/evidence#2372] [General]
A waterfall chart is great for showing decomposition of changes.

I've had a start at creating one (which I will open in a draft PR for those who are interested in taking a look).
<img width="699" alt="CleanShot 2023-04-26 at 10 42 45@2x" src="https://user-images.githubusercontent.com/12602440/234636102-dbd43f3d-8d11-4d5b-950d-32f745dcc805.png">

<img width="597" alt="CleanShot 2023-04-26 at 10 57 15@2x" src="https://user-images.githubusercontent.com/12602440/234636122-e94cbd04-926e-40b9-a21b-2c6a66de0c44.png">

## Considerations for implementation
- Axis labels are important on these charts and will often be long strings - the chart needs to be able to handle multi-line labels without cutting off characters or breaking words apart
   - I think it probably needs "breaking" behaviour by word, followed by either truncation (`...`) or expansion of the chart size
   - This will also need to work well responsively and on mobile
- We don't currently have multi-line axis labels in our other charts - this will need some thinking to make sure the chart area is the right size in all cases. Currently, adding multi-line axis labels shrinks the main chart area
- The initial attempt I've made requires a specific data format: 
   - `category` - category to use on x-axis
   - `value` - change to show on chart
   - `total` - boolean - should this be treated as a total/subtotal on the chart
- The SQL required to get data into this format can require a lot of work. We might want to offer other alternatives for the structure of the data getting passed in
   - I don't think this would change the implementation within echarts - we could just add a layer of logic above this to get it into the format echarts requires
- Data labels should be included - positioning these may be difficult   
- I haven't thought about tooltips yet. I'm thinking they will be straightforward though - probably a config override sent from `Waterfall` to `Chart`
- Subtotals - the logic for generating the chart series will need to be updated to accomodate subtotals - currently, having a total in the middle of the data throws off the cumulative sum
- We may want to offer sorting - e.g, sort by magnitude of change. This shouldn't be hard to do with this data format since we can easily identify the totals and exclude them from sorting
- In a lot of cases, people need to start the y-axis at a higher number than 0 to make the changes visible enough on a waterfall chart
   - We could offer a `yMin` prop to allow someone to set it manually
   - I could also imagine some dynamic logic to check the change range (max of cumulative sum - min of cumulative sum) as a % of axis range and adjust the `yMin` so that the change range takes up something like 50% of the vertical chart area

**Comment by matsonj** [comment]
> The SQL required to get data into this format can require a lot of work. We might want to offer other alternatives for the structure of the data getting passed in

I think including a sample SQL query in the documentation would go a long way. It is indeed fairly complex but you can simplify it massively with clever windowing (LAG / LEAD or cumulative sums). I think having a clean example of the data model on this one is key because it always takes me like an hour in power BI to get these right.

**Comment by archiewood** [comment]
@matsonj Does the below interpret what you are saying correctly?

There are a few common uses for waterfalls, for example
- Revenue bridge (by category, geography etc)
- Volume value bridge (revenue start, impact of price, impact of volume, impact of mix, residual, revenue end) 

We should write an example of how to wrangle a standard table format in order to do a waterfall of the above types

eg

Assuming your columns are this:
- `date`, `revenue`, `category`
- Then use this example query to get into waterfall format

Assuming your columns are this
- `date`, `item`, `price`, `quantity`, [total revenue]
- Then use this other query

And as a follow up question: what are your most typical waterfall use cases (assuming not covered by the above two)?

**Comment by matsonj** [comment]
@archiewood yeah, I think those are the core cases. I think the core questions I usually run into that usually are answered best with waterfalls are the following:

- Why did revenue change this month vs last in percentage and absolute $? (i.e. change in revenue by type - recurring, churn, growth, etc)
- Why did profitability change this month vs last month in percentage and absolute $? (i.e. contribution margin by product line)

Could go down the line on any GL line item too, but those are the two that seem the most frequent in my exp.

**Comment by rasmusengelbrechtsorensen** [comment]
Is Waterfall still on the Backlog, or is it closed for now? 😊

**Comment by hughess** [comment]
Still on the backlog, would definitely welcome contributions!

**Reply by rasmusengelbrechtsorensen** [reply]
Awesome! Unfortunately, I don't think I have the abilities to contribute... 😅
